Sources/BrowserServicesKit/ReferrerTrimming/ReferrerTrimming.swift

```diff
@@ -0,0 +1,144 @@
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import WebKit
+import TrackerRadarKit
+
+public class ReferrerTrimming {
+    
+    struct Constants {
+        static let headerName = "Referer"
+        static let policyName = "Referrer-Policy"
+    }
+    
+    public enum TrimmingState {
+        case idle
+        case navigating(destination: URL)
+    }
+    
+    private let privacyManager: PrivacyConfigurationManager
+    private var privacyConfig: PrivacyConfiguration { privacyManager.privacyConfig }
+    
+    private let contentBlockingManager: ContentBlockerRulesManager
+    
+    private var state: TrimmingState = .idle
+    
+    private var tld: TLD
+    
+    public init(privacyManager: PrivacyConfigurationManager,
+                contentBlockingManager: ContentBlockerRulesManager,
+                tld: TLD) {
+        self.privacyManager = privacyManager
+        self.contentBlockingManager = contentBlockingManager
+        self.tld = tld
+    }
+    
+    public func onBeginNavigation(to destination: URL?) {
+        guard let destination = destination else {
+            return
+        }
+        
+        state = .navigating(destination: destination)
+    }
+    
+    public func onFinishNavigation() {
+        state = .idle
+    }
+    
+    public func onFailedNavigation() {
+        state = .idle
+    }
+    
+    func getTrimmedReferrer(originUrl: URL, destUrl: URL, referrerUrl: URL?, trackerData: TrackerData) -> String? {
+        func isSameEntity(a: Entity?, b: Entity?) -> Bool {
+            if a == nil && b == nil {
+                return !originUrl.isThirdParty(to: destUrl, tld: tld)
+            }
+            
+            return a?.displayName == b?.displayName
+        }
+        
+        guard privacyConfig.isFeature(.referrer, enabledForDomain: originUrl.host!),
+              privacyConfig.isFeature(.referrer, enabledForDomain: destUrl.host!) else {
+            return nil
+        }
+        guard let referrerUrl = referrerUrl else {
+            return nil
+        }
+        
+        guard originUrl.host != nil, originUrl.scheme != nil else {
+            return nil
+        }
+        guard destUrl.host != nil, destUrl.scheme != nil else {
+            return nil
+        }
+        
+        let referEntity = trackerData.findEntity(forHost: originUrl.host!)
+        let destEntity = trackerData.findEntity(forHost: destUrl.host!)
```

Comment:
Same here - better pattern is to assign host to variable in guard rather than simply check it there.